File: models/Gpt-3.5-turbo/py/code/string_algorithms/String_Matching.py

def compute_lps(pattern):
    lps = [0] * len(pattern)
    length = 0  # length of the previous longest prefix suffix

    i = 1
    while i < len(pattern):
        if pattern[i] == pattern[length]:
            length += 1
            lps[i] = length
            i += 1
        else:
            if length != 0:
                # Try to find a smaller prefix that is also a suffix
                length = lps[length - 1]
            else:
                lps[i] = 0
                i += 1

    return lps


def count_occurrences(string, pattern):
    n = len(string)
    m = len(pattern)

    lps = compute_lps(pattern)

    i = 0  # Index for string
    j = 0  # Index for pattern
    count = 0  # Number of occurrences

    while i < n:
        if string[i] == pattern[j]:
            i += 1
            j += 1

            if j == m:
                # Pattern found
                count += 1
                j = lps[j - 1]

        else:
            if j != 0:
                # Try to find a smaller prefix that is also a suffix
                j = lps[j - 1]
            else:
                i += 1

    return count


# Example usage:
string = input()
pattern = input()

result = count_occurrences(string, pattern)
print(result)

File: models/Gpt-3.5-turbo/cpp/code/problems/string_algorithms/String_Matching.cpp

#include <iostream>
#include <string>
using namespace std;

int countOccurrences(string s, string pattern) {
    int n = s.length();
    int m = pattern.length();
    int count = 0;

    for (int i = 0; i <= n - m; i++) {
        if (s.substr(i, m) == pattern) {
            count++;
        }
    }

    return count;
}

int main() {
    string s, pattern;
    cin >> s >> pattern;

    int occurrences = countOccurrences(s, pattern);
    cout << occurrences << endl;

    return 0;
}
